// Copyright 2018 The Paddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// =============================================================================

syntax = "proto3";
option optimize_for = LITE_RUNTIME;

package paddle_hub;

enum DataType {
  INT = 0;
  FLOAT = 1;
  STRING = 2;
  BOOLEAN = 3;
  LIST = 4;
  MAP = 5;
}

message FlexibleData {
  DataType type = 1;
  string name = 2;
  int32 i = 3;
  float f = 4;
  bool b = 5;
  string s = 6;
  map<string, FlexibleData> m = 7;
  map<int32, FlexibleData> l = 8;
}

// Feed Variable Description
message FeedDesc {
  string var_name = 1;
  string alias = 2;
};

// Fetch Variable Description
message FetchDesc {
  string var_name = 1;
  string alias = 2;
};

// Module Variable
message ModuleVar {
  repeated FetchDesc fetch_desc = 1;
  repeated FeedDesc feed_desc = 2;
}

message AuthInfo {
  string paddle_version = 1;
  string hub_version = 2;
}

message ParamAttr {
  message Regularizer {
    string type = 1;
    float regularization_coeff = 2;
  }
  
  message GradientClipAttr {
    string type = 1;
    float min = 2;
    float max = 3;
    float clip_norm = 4;
    string group_name = 5;
  }
  
  Regularizer regularizer = 1;
  GradientClipAttr gradient_clip_attr = 2;
  FlexibleData optimize_attr = 3;
  bool trainable = 4;
  bool do_model_average = 5;
}

// A Hub Module is stored in a directory with a file 'paddlehub.pb'
// containing a serialized protocol message of this type. The further contents
// of the directory depend on the storage format described by the message.
message ModuleDesc {
  // PaddleHub module name
  string name = 1; // PaddleHub module name
  
  // signature to module variable
  map<string, ModuleVar> sign2var = 2;

  bool return_numpy = 3;

  bool contain_assets = 4;

  AuthInfo auth_info = 5;
  
  map<string, ParamAttr> param_attrs = 6;
};

